functional/ldap_test.go

```diff
@@ -123,3 +125,52 @@ func TestConnectorLDAPHealthy(t *testing.T) {
 		}
 	}
 }
+
+func TestLDAPPool(t *testing.T) {
+	server := ldapServer(t)
+	ldapPool := &connector.LDAPPool{
+		MaxIdleConn: 30,
+		ServerHost:  server.Host,
+		ServerPort:  server.Port,
+		UseTLS:      false,
+		UseSSL:      false,
+	}
+
+	// Excercise pool operations with MaxIdleConn + 10 concurrent goroutines.
+	// The 10 surplus connections should be discarded after use.
+	// Remaining active connections will be verified at the end of the test.
+	numRoutines := ldapPool.MaxIdleConn + 10
+	var wg sync.WaitGroup
+	for i := 0; i < numRoutines; i++ {
+		wg.Add(1)
+		go func() {
+			timeout := time.After(5 * time.Second)
+			defer wg.Done()
+			for {
+				select {
+				case <-timeout:
+					return
+				default:
+					ldapConn, err := ldapPool.Acquire()
+					if err != nil {
+						t.Errorf("Unable to acquire LDAP Connection: %v", err)
+					}
+					s := ldap.NewSearchRequest("", ldap.ScopeBaseObject, ldap.NeverDerefAliases, 0, 0, false, "(objectClass=*)", []string{}, nil)
+					_, err = ldapConn.Search(s)
+					if err != nil {
+						t.Errorf("Search request failed. Dead/invalid LDAP connection from pool?: %v", err)
+					}
+					ldapPool.Put(ldapConn)
+					_, _ = ldapPool.CheckConnections()
+				}
+			}
+		}()
+	}
+
+	// Wait for all operations to complete and check status.
+	wg.Wait()
+	alive, killed := ldapPool.CheckConnections()
```

Comment:
Checking that there are MaxIdleConn connections active after the test gives us information about two things:
1. Did we indeed have concurrent operations during the test.
2. The test runs more than MaxIdleConn connections concurrently, and checking that the pool stays at MaxIdleConn after the run confirms that the code handling the high watermark of the pool works.